io/io/src/RFile.cxx

```diff
@@ -0,0 +1,381 @@
+TKey *RFile::GetTKey(std::string_view path) const
+{
+   // In RFile, differently from TFile, when dealing with a path like "a/b/c", we always consider it to mean
+   // "object 'c' in subdirectory 'b' of directory 'a'". We don't try to get any other of the possible combinations,
+   // including the object called "a/b/c".
+   std::string fullPath = std::string(path);
+   char *dirName = fullPath.data();
+   char *restOfPath = strchr(dirName, '/');
+   TDirectory *dir = fFile.get();
+   while (restOfPath) {
+      // Truncate `dirName` to the position of this '/'.
+      *restOfPath = 0;
+      ++restOfPath;
+      // `restOfPath` should always be a non-empty string unless `path` ends with '/' (which it shouldn't, as we are
+      // supposed to have normalized it before calling this function).
+      assert(*restOfPath);
+
+      dir = dir->GetDirectory(dirName);
+      if (!dir)
+         return nullptr;
+
+      dirName = restOfPath;
+      restOfPath = strchr(restOfPath, '/');
+   }
+   // NOTE: after this loop `dirName` contains the base name of the object.
+
+   // Get the leaf object from the innermost directory.
+   TKey *key = dir->FindKey(dirName);
+   if (key) {
+      // For some reason, FindKey will not return nullptr if we asked for a specific cycle and that cycle
+      // doesn't exist. It will instead return any key whose cycle is *at most* the requested one.
+      // This is very confusing, so in RFile we actually return null if the requested cycle is not there.
+      RNameCycleResult res = DecodeNameCycle(dirName);
+      if (res.fError != ENameCycleError::kAnyCycle) {
+         if (res.fError != ENameCycleError::kNoError) {
+            R__LOG_ERROR(RFileLog()) << "error decoding namecycle '" << dirName << "': " << ToString(res.fError);
+            key = nullptr;
+         } else if (res.fCycle && *res.fCycle != key->GetCycle()) {
+            key = nullptr;
+         }
+      }
+   }
+   return key;
+}
+
+void *RFile::GetUntyped(std::string_view pathSV, const std::type_info &type) const
+{
+   if (!fFile)
+      throw ROOT::RException(R__FAIL("File has been closed"));
+
+   std::string path{pathSV};
+
+   const TClass *cls = TClass::GetClass(type);
+   if (!cls)
+      throw ROOT::RException(R__FAIL(std::string("Could not determine type of object ") + path));
+
+   if (auto err = ValidateAndNormalizePath(path); !err.empty())
+      throw RException(R__FAIL("Invalid object path '" + path + "': " + err));
+
+   TKey *key = GetTKey(path);
+   void *obj = key ? key->ReadObjectAny(cls) : nullptr;
+
+   if (obj) {
+      // Disavow any ownership on `obj`
+      if (auto autoAddFunc = cls->GetDirectoryAutoAdd(); autoAddFunc) {
+         autoAddFunc(obj, nullptr);
+      }
+   } else if (key && !GetROOT()->IsBatch()) {
+      R__LOG_WARNING(RFileLog()) << "Tried to get object '" << path << "' of type " << cls->GetName()
+                                 << " but that path contains an object of type " << key->GetClassName();
+   }
```

Comment:
This is debatable.   What if the user wants to use this in a script used interactively?  For example, the inner part of the script might be interrogating what 'user type' of file this before moving forward (eg. do something for file containing `TTree` with one name and another for files containing a `RNTuple` with a different name.